// @wundergraph/sdk/dist/db/introspection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntrospectMySQLDatabase = exports.IntrospectPostgresDatabase = void 0;
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const child_process_1 = require("child_process");
const introspectPrismaDatabase = (queryString, databaseSchema) => {
    var _a;
    const home = os_1.default.homedir();
    const cmd = path_1.default.join(home, ".wundergraph", "wunderctl");
    const result = child_process_1.spawnSync(cmd, ["introspect", databaseSchema, queryString], {
        cwd: process.cwd(),
        encoding: "utf-8",
        timeout: 10000,
    });
    if (result.status !== 0) {
        return {
            success: false,
            message: (_a = result.error) === null || _a === void 0 ? void 0 : _a.message,
            graphql_schema: "",
            prisma_schema: "",
        };
    }
    const output = result.output.join("");
    const out = JSON.parse(output);
    out.success = true;
    return out;
};
const IntrospectPostgresDatabase = (queryString) => {
    return introspectPrismaDatabase(queryString, "postgresql");
};
exports.IntrospectPostgresDatabase = IntrospectPostgresDatabase;
const IntrospectMySQLDatabase = (queryString) => {
    return introspectPrismaDatabase(queryString, "mysql");
};
exports.IntrospectMySQLDatabase = IntrospectMySQLDatabase;
