// @wundergraph/sdk/dist/graphql/operations.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const operations_1 = require("./operations");
const chai_1 = require("chai");
const graphql_1 = require("graphql");
const wundernode_config_1 = require("@wundergraph/protobuf/dist/wundernode_config");
const fs = __importStar(require("fs"));
const path_1 = __importDefault(require("path"));
const operations = `
query MyReviews {
    me {
        name
        reviews {
            id
            body
        }
    }
}

mutation CreatePet ($petInput: PetInput!) {
    postPets(petInput: $petInput){
        name
    }
}

subscription NewPets {
    newPets {
        name
    }
}`;
const expected = {
    operations: [
        {
            Name: "MyReviews",
            Content: "query MyReviews {\n  me {\n    name\n    reviews {\n      id\n      body\n    }\n  }\n}",
            OperationType: wundernode_config_1.OperationType.QUERY,
            VariablesSchema: {
                type: "object",
                properties: {},
                additionalProperties: false,
            },
            ResponseSchema: {
                type: "object",
                properties: {
                    "data": {
                        type: "object",
                        properties: {
                            "me": {
                                type: "object",
                                properties: {
                                    "name": {
                                        type: "string",
                                    },
                                    "reviews": {
                                        type: "array",
                                        items: {
                                            type: "object",
                                            properties: {
                                                "id": {
                                                    type: "string",
                                                },
                                                "body": {
                                                    type: "string"
                                                }
                                            },
                                            additionalProperties: false,
                                            required: [
                                                "id"
                                            ]
                                        }
                                    }
                                },
                                additionalProperties: false,
                            }
                        },
                        additionalProperties: false,
                    },
                },
                additionalProperties: false,
            },
            AuthenticationConfig: {
                required: false,
            },
            AuthorizationConfig: {
                claims: [],
            }
        },
        {
            Name: "CreatePet",
            Content: "mutation CreatePet($petInput: PetInput!) {\n  postPets(petInput: $petInput) {\n    name\n  }\n}",
            OperationType: wundernode_config_1.OperationType.MUTATION,
            VariablesSchema: {
                type: "object",
                properties: {
                    "petInput": {
                        additionalProperties: false,
                        type: "object",
                        properties: {
                            "id": {
                                type: "string"
                            },
                            "name": {
                                type: "string"
                            }
                        },
                        required: [
                            "id",
                        ]
                    },
                },
                additionalProperties: false,
                required: [
                    "petInput",
                ],
            },
            ResponseSchema: {
                type: "object",
                properties: {
                    "data": {
                        type: "object",
                        properties: {
                            "postPets": {
                                type: "object",
                                properties: {
                                    "name": {
                                        type: "string"
                                    }
                                },
                                additionalProperties: false,
                            }
                        },
                        additionalProperties: false,
                    },
                },
                additionalProperties: false,
            },
            AuthenticationConfig: {
                required: false,
            },
            AuthorizationConfig: {
                claims: [],
            }
        },
        {
            Name: "NewPets",
            Content: "subscription NewPets {\n  newPets {\n    name\n  }\n}",
            OperationType: wundernode_config_1.OperationType.SUBSCRIPTION,
            VariablesSchema: {
                type: "object",
                properties: {},
                additionalProperties: false,
            },
            ResponseSchema: {
                type: "object",
                properties: {
                    "data": {
                        type: "object",
                        properties: {
                            "newPets": {
                                type: "object",
                                properties: {
                                    "name": {
                                        type: "string"
                                    }
                                },
                                additionalProperties: false,
                            }
                        },
                        additionalProperties: false,
                    },
                },
                additionalProperties: false,
            },
            AuthenticationConfig: {
                required: false,
            },
            AuthorizationConfig: {
                claims: [],
            }
        },
    ]
};
test("parseGraphQLOperations", () => {
    const actual = operations_1.parseOperations(testSchema, operations);
    chai_1.assert.equal(pretty(actual), pretty(expected));
});
const fromClaimSchema = `
schema {
    mutation: Mutation
}
type Mutation {
    createUser(email: String!): User
}
type User {
    email: String!
}
directive @fromClaim(name: Claim) on VARIABLE_DEFINITION
enum Claim {
  EMAIL
  EMAIL_VERIFIED
  NAME
  NICKNAME
  LOCATION
  PROVIDER
}
`;
const fromClaimOperations = `
mutation CreateUser(
    $email: String! @fromClaim(name: EMAIL)
) {
    createUser(email: $email) {
        email
    }
}
`;
const fromClaimParsed = {
    operations: [
        {
            Name: "CreateUser",
            Content: "mutation CreateUser($email: String! @fromClaim(name: EMAIL)) {\n  createUser(email: $email) {\n    email\n  }\n}",
            OperationType: wundernode_config_1.OperationType.MUTATION,
            VariablesSchema: {
                type: "object",
                properties: {},
                additionalProperties: false,
            },
            ResponseSchema: {
                type: "object",
                properties: {
                    "data": {
                        type: "object",
                        properties: {
                            "createUser": {
                                type: "object",
                                properties: {
                                    "email": {
                                        type: "string",
                                    }
                                },
                                additionalProperties: false,
                                required: ["email"],
                            }
                        },
                        additionalProperties: false,
                    }
                },
                additionalProperties: false,
            },
            AuthenticationConfig: {
                required: true,
            },
            AuthorizationConfig: {
                claims: [
                    {
                        variableName: "email",
                        claim: wundernode_config_1.Claim.EMAIL,
                    }
                ]
            }
        }
    ]
};
test("parse operations with fromClaim diretive", () => {
    const actual = operations_1.parseOperations(fromClaimSchema, fromClaimOperations);
    chai_1.assert.equal(pretty(actual), pretty(fromClaimParsed));
});
const testSchema = `type Query {
  me: User
  topProducts(first: Int = 5): [Product]
  nullableComplex(input: ComplexInput): User
  nonNullComplex(input: ComplexInput!): User
  nonNullMixed(input: ComplexInput!, a: Int!): User
}

type Mutation {
    postPets(petInput: PetInput!): Pet
}

type Subscription {
    newPets: Pet
}

type Pet {
    id: ID!
    name: String
}

input PetInput {
    id: ID!
    name: String
}

enum EnumInput {
    A
    B
}

input ComplexInput {
    nullableString: String
    nullableStringWithDefault: String = "foo"
    nonNullString: String!
    nonNullStringWithDefault: String! = "bar"
    nullableBoolean: Boolean
    nonNullBoolean: Boolean!
    nullableFloat: Float
    nonNullFloat: Float!
    nullableInt: Int
    nonNullInt: Int!
    nullableInput: NestedInput
    nonNullInput: NestedInput!
    nullableEnum: EnumInput
    nonNullEnum: EnumInput!
    nullableListOfNullableString: [String]
    nonNullListOfNullableString: [String]!
    nonNullListOfNonNullString: [String!]!
}

input NestedInput {
    nullableString: String
}

type User {
  id: ID!
  name: String
  username: String
  reviews: [Review]
}

type Product {
  upc: String!
  name: String
  price: Int
  weight: Int
  reviews: [Review]
  inStock: Boolean
  shippingEstimate: Int
}

type Review {
  id: ID!
  body: String
  author: User
  product: Product
}
`;
test("operationVariablesToJSONSchema", () => {
    subTest(testSchema, [
        {
            query: `query Q($a: Int){topProducts(first: $a){upc}}`,
            expected: {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "object",
                        "properties": {
                            "topProducts": {
                                "type": "array",
                                "items": {
                                    "type": "object",
                                    "properties": {
                                        "upc": {
                                            "type": "string"
                                        }
                                    },
                                    "additionalProperties": false,
                                    "required": [
                                        "upc"
                                    ]
                                }
                            }
                        },
                        "additionalProperties": false
                    }
                },
                "additionalProperties": false
            },
            expectedVariables: {
                type: "object",
                properties: {
                    "a": {
                        type: "integer",
                    }
                },
                additionalProperties: false,
            }
        },
        {
            query: `query Q($a: ComplexInput){nullableComplex(input: $a){id}}`,
            expected: {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "object",
                        "properties": {
                            "nullableComplex": {
                                "type": "object",
                                "properties": {
                                    "id": {
                                        "type": "string"
                                    }
                                },
                                "additionalProperties": false,
                                "required": [
                                    "id"
                                ]
                            }
                        },
                        "additionalProperties": false
                    }
                },
                "additionalProperties": false
            },
            expectedVariables: {
                type: "object",
                properties: {
                    "a": {
                        additionalProperties: false,
                        type: "object",
                        properties: {
                            "nullableString": {
                                type: 'string'
                            },
                            "nullableStringWithDefault": {
                                type: "string",
                            },
                            "nonNullString": {
                                type: "string",
                            },
                            "nonNullStringWithDefault": {
                                type: "string",
                            },
                            "nullableBoolean": {
                                type: "boolean"
                            },
                            "nonNullBoolean": {
                                type: "boolean",
                            },
                            "nullableFloat": {
                                type: "number",
                            },
                            "nonNullFloat": {
                                type: "number",
                            },
                            "nullableInt": {
                                type: "integer",
                            },
                            "nonNullInt": {
                                type: "integer",
                            },
                            "nullableInput": {
                                additionalProperties: false,
                                type: "object",
                                properties: {
                                    "nullableString": {
                                        type: "string",
                                    }
                                }
                            },
                            "nonNullInput": {
                                additionalProperties: false,
                                type: "object",
                                properties: {
                                    "nullableString": {
                                        type: "string",
                                    }
                                }
                            },
                            "nullableEnum": {
                                type: "string",
                                enum: ["A", "B"]
                            },
                            "nonNullEnum": {
                                type: "string",
                                enum: ["A", "B"]
                            },
                            "nullableListOfNullableString": {
                                type: "array",
                                items: {
                                    type: "string"
                                }
                            },
                            "nonNullListOfNullableString": {
                                type: "array",
                                items: {
                                    type: "string"
                                }
                            },
                            "nonNullListOfNonNullString": {
                                type: "array",
                                minItems: 1,
                                items: {
                                    type: "string"
                                }
                            },
                        },
                        required: [
                            "nonNullString",
                            "nonNullBoolean",
                            "nonNullFloat",
                            "nonNullInt",
                            "nonNullInput",
                            "nonNullEnum",
                            "nonNullListOfNullableString",
                            "nonNullListOfNonNullString"
                        ],
                    },
                },
                additionalProperties: false,
            }
        }
    ]);
});
const pretty = (input) => {
    return JSON.stringify(input, null, "  ");
};
const subTest = (rawSchema, testCase) => {
    const schema = graphql_1.buildSchema(rawSchema);
    testCase.forEach((testCase, i) => {
        const queryDocument = graphql_1.parse(testCase.query);
        const operation = queryDocument.definitions.find(node => node.kind === "OperationDefinition");
        const actual = operations_1.operationResponseToJSONSchema(schema, queryDocument, operation);
        chai_1.assert.equal(pretty(actual), pretty(testCase.expected), "testCase: " + i);
        if (testCase.expectedVariables === undefined) {
            return;
        }
        const actualVariables = operations_1.operationVariablesToJSONSchema(schema, operation);
        chai_1.assert.equal(pretty(actualVariables), pretty(testCase.expectedVariables), "testCase: " + i);
    });
};
test("complex schema operations", () => {
    subTest(complexSchema, [
        {
            query: `mutation CreateNamespace($name: String! $personal: Boolean!) {
                        namespaceCreate(input: {name: $name, personal: $personal}){
                            ... on NamespaceCreated {
                                namespace {
                                    id
                                    name
                                }
                            }
                            ... on Error {
                                code
                                message
                            }
                        }
                    }`,
            expected: {
                type: "object",
                properties: {
                    "data": {
                        type: "object",
                        properties: {
                            "namespaceCreate": {
                                type: "object",
                                properties: {
                                    "namespace": {
                                        type: "object",
                                        properties: {
                                            "id": {
                                                type: "string",
                                            },
                                            "name": {
                                                type: "string",
                                            }
                                        },
                                        additionalProperties: false,
                                        required: [
                                            "id",
                                            "name",
                                        ]
                                    },
                                    "code": {
                                        type: "string",
                                        enum: [
                                            "Internal",
                                            "AuthenticationRequired",
                                            "Unauthorized",
                                            "NotFound",
                                            "Conflict",
                                            "UserAlreadyHasPersonalNamespace",
                                            "TeamPlanInPersonalNamespace",
                                            "InvalidName",
                                            "UnableToDeployEnvironment",
                                            "InvalidWunderGraphConfig",
                                            "ApiEnvironmentNamespaceMismatch",
                                            "UnableToUpdateEdgesOnPersonalEnvironment",
                                        ]
                                    },
                                    "message": {
                                        type: "string",
                                    }
                                },
                                additionalProperties: false,
                            }
                        },
                        additionalProperties: false,
                        required: [
                            "namespaceCreate"
                        ],
                    }
                },
                additionalProperties: false,
            }
        }
    ]);
});
test("complex operations with variables", () => {
    const rawSchema = fs.readFileSync(path_1.default.resolve(__dirname, "./testdata/wundergraph.schema.graphql.file"));
    const schema = rawSchema.toString("utf-8");
    subTest(schema, [
        {
            query: `
            query FirstAPI($id: String!) {
                findFirstapi(where: {id: {equals: $id}}) {
                    id
                    name
                    namespace {
                        id
                        name
                    }
                }
            }`,
            expected: {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "object",
                        "properties": {
                            "findFirstapi": {
                                "type": "object",
                                "properties": {
                                    "id": {
                                        "type": "string"
                                    },
                                    "name": {
                                        "type": "string"
                                    },
                                    "namespace": {
                                        "type": "object",
                                        "properties": {
                                            "id": {
                                                "type": "string"
                                            },
                                            "name": {
                                                "type": "string"
                                            }
                                        },
                                        "additionalProperties": false,
                                        "required": [
                                            "id",
                                            "name"
                                        ]
                                    }
                                },
                                "additionalProperties": false,
                                "required": [
                                    "id",
                                    "name",
                                    "namespace"
                                ]
                            }
                        },
                        "additionalProperties": false
                    }
                },
                "additionalProperties": false
            },
            expectedVariables: {
                "type": "object",
                "properties": {
                    "id": {
                        "type": "string"
                    }
                },
                "additionalProperties": false,
                "required": [
                    "id"
                ]
            }
        },
        {
            query: `
            query FirstNamespace($id: StringFilter!) {
                findFirstnamespace(where: {id: $id}) {
                    id
                    name
                    api {
                        id
                        name
                        created_at
                    }
                }
            }`,
            expected: {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "object",
                        "properties": {
                            "findFirstnamespace": {
                                "type": "object",
                                "properties": {
                                    "id": {
                                        "type": "string"
                                    },
                                    "name": {
                                        "type": "string"
                                    },
                                    "api": {
                                        "type": "array",
                                        "items": {
                                            "type": "object",
                                            "properties": {
                                                "id": {
                                                    "type": "string"
                                                },
                                                "name": {
                                                    "type": "string"
                                                },
                                                "created_at": {
                                                    "type": "string"
                                                }
                                            },
                                            "additionalProperties": false,
                                            "required": [
                                                "id",
                                                "name",
                                                "created_at"
                                            ]
                                        }
                                    }
                                },
                                "additionalProperties": false,
                                "required": [
                                    "id",
                                    "name"
                                ]
                            }
                        },
                        "additionalProperties": false
                    }
                },
                "additionalProperties": false
            },
            expectedVariables: {
                "type": "object",
                "properties": {
                    "id": {
                        "additionalProperties": false,
                        "type": "object",
                        "properties": {
                            "equals": {
                                "type": "string"
                            },
                            "in": {
                                "type": "array",
                                "items": {
                                    "type": "string"
                                }
                            },
                            "notIn": {
                                "type": "array",
                                "items": {
                                    "type": "string"
                                }
                            },
                            "lt": {
                                "type": "string"
                            },
                            "lte": {
                                "type": "string"
                            },
                            "gt": {
                                "type": "string"
                            },
                            "gte": {
                                "type": "string"
                            },
                            "contains": {
                                "type": "string"
                            },
                            "startsWith": {
                                "type": "string"
                            },
                            "endsWith": {
                                "type": "string"
                            },
                            "mode": {
                                "type": "string",
                                "enum": [
                                    "default",
                                    "insensitive"
                                ]
                            },
                            "not": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "equals": {
                                        "type": "string"
                                    },
                                    "in": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "notIn": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "lt": {
                                        "type": "string"
                                    },
                                    "lte": {
                                        "type": "string"
                                    },
                                    "gt": {
                                        "type": "string"
                                    },
                                    "gte": {
                                        "type": "string"
                                    },
                                    "contains": {
                                        "type": "string"
                                    },
                                    "startsWith": {
                                        "type": "string"
                                    },
                                    "endsWith": {
                                        "type": "string"
                                    },
                                    "not": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "not": {}
                                        }
                                    }
                                }
                            }
                        }
                    }
                },
                "additionalProperties": false,
                "required": [
                    "id"
                ]
            }
        },
        {
            query: `
            query FirstNamespace($where: namespaceWhereInput) {
                findFirstnamespace(where: $where) {
                    id
                    name
                    api {
                        id
                        name
                        created_at
                    }
                }
            }`, expected: {
                "type": "object",
                "properties": {
                    "data": {
                        "type": "object",
                        "properties": {
                            "findFirstnamespace": {
                                "type": "object",
                                "properties": {
                                    "id": {
                                        "type": "string"
                                    },
                                    "name": {
                                        "type": "string"
                                    },
                                    "api": {
                                        "type": "array",
                                        "items": {
                                            "type": "object",
                                            "properties": {
                                                "id": {
                                                    "type": "string"
                                                },
                                                "name": {
                                                    "type": "string"
                                                },
                                                "created_at": {
                                                    "type": "string"
                                                }
                                            },
                                            "additionalProperties": false,
                                            "required": [
                                                "id",
                                                "name",
                                                "created_at"
                                            ]
                                        }
                                    }
                                },
                                "additionalProperties": false,
                                "required": [
                                    "id",
                                    "name"
                                ]
                            }
                        },
                        "additionalProperties": false
                    }
                },
                "additionalProperties": false
            },
            expectedVariables: {
                "type": "object",
                "properties": {
                    "where": {
                        "additionalProperties": false,
                        "type": "object",
                        "properties": {
                            "AND": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "AND": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "price_plan_id": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "price_plan": {},
                                            "api": {},
                                            "environment": {},
                                            "namespace_members": {}
                                        }
                                    },
                                    "OR": {
                                        "type": "array",
                                        "items": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "AND": {},
                                                "OR": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "NOT": {},
                                                "id": {},
                                                "name": {},
                                                "price_plan_id": {},
                                                "created_at": {},
                                                "updated_at": {},
                                                "price_plan": {},
                                                "api": {},
                                                "environment": {},
                                                "namespace_members": {}
                                            }
                                        }
                                    },
                                    "NOT": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "price_plan_id": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "price_plan": {},
                                            "api": {},
                                            "environment": {},
                                            "namespace_members": {}
                                        }
                                    },
                                    "id": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "mode": {},
                                            "not": {}
                                        }
                                    },
                                    "name": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "mode": {},
                                            "not": {}
                                        }
                                    },
                                    "price_plan_id": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "integer"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "integer"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "integer"
                                                }
                                            },
                                            "lt": {
                                                "type": "integer"
                                            },
                                            "lte": {
                                                "type": "integer"
                                            },
                                            "gt": {
                                                "type": "integer"
                                            },
                                            "gte": {
                                                "type": "integer"
                                            },
                                            "not": {}
                                        }
                                    },
                                    "created_at": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {},
                                            "in": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "lt": {},
                                            "lte": {},
                                            "gt": {},
                                            "gte": {},
                                            "not": {}
                                        }
                                    },
                                    "updated_at": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {},
                                            "in": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "lt": {},
                                            "lte": {},
                                            "gt": {},
                                            "gte": {},
                                            "not": {}
                                        }
                                    },
                                    "price_plan": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "is": {},
                                            "isNot": {}
                                        }
                                    },
                                    "api": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "every": {},
                                            "some": {},
                                            "none": {}
                                        }
                                    },
                                    "environment": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "every": {},
                                            "some": {},
                                            "none": {}
                                        }
                                    },
                                    "namespace_members": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "every": {},
                                            "some": {},
                                            "none": {}
                                        }
                                    }
                                }
                            },
                            "OR": {
                                "type": "array",
                                "items": {
                                    "additionalProperties": false,
                                    "type": "object",
                                    "properties": {
                                        "AND": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "AND": {},
                                                "OR": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "NOT": {},
                                                "id": {},
                                                "name": {},
                                                "price_plan_id": {},
                                                "created_at": {},
                                                "updated_at": {},
                                                "price_plan": {},
                                                "api": {},
                                                "environment": {},
                                                "namespace_members": {}
                                            }
                                        },
                                        "OR": {
                                            "type": "array",
                                            "items": {
                                                "additionalProperties": false,
                                                "type": "object",
                                                "properties": {
                                                    "AND": {},
                                                    "OR": {
                                                        "type": "array",
                                                        "items": {}
                                                    },
                                                    "NOT": {},
                                                    "id": {},
                                                    "name": {},
                                                    "price_plan_id": {},
                                                    "created_at": {},
                                                    "updated_at": {},
                                                    "price_plan": {},
                                                    "api": {},
                                                    "environment": {},
                                                    "namespace_members": {}
                                                }
                                            }
                                        },
                                        "NOT": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "AND": {},
                                                "OR": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "NOT": {},
                                                "id": {},
                                                "name": {},
                                                "price_plan_id": {},
                                                "created_at": {},
                                                "updated_at": {},
                                                "price_plan": {},
                                                "api": {},
                                                "environment": {},
                                                "namespace_members": {}
                                            }
                                        },
                                        "id": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "equals": {
                                                    "type": "string"
                                                },
                                                "in": {
                                                    "type": "array",
                                                    "items": {
                                                        "type": "string"
                                                    }
                                                },
                                                "notIn": {
                                                    "type": "array",
                                                    "items": {
                                                        "type": "string"
                                                    }
                                                },
                                                "lt": {
                                                    "type": "string"
                                                },
                                                "lte": {
                                                    "type": "string"
                                                },
                                                "gt": {
                                                    "type": "string"
                                                },
                                                "gte": {
                                                    "type": "string"
                                                },
                                                "contains": {
                                                    "type": "string"
                                                },
                                                "startsWith": {
                                                    "type": "string"
                                                },
                                                "endsWith": {
                                                    "type": "string"
                                                },
                                                "mode": {},
                                                "not": {}
                                            }
                                        },
                                        "name": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "equals": {
                                                    "type": "string"
                                                },
                                                "in": {
                                                    "type": "array",
                                                    "items": {
                                                        "type": "string"
                                                    }
                                                },
                                                "notIn": {
                                                    "type": "array",
                                                    "items": {
                                                        "type": "string"
                                                    }
                                                },
                                                "lt": {
                                                    "type": "string"
                                                },
                                                "lte": {
                                                    "type": "string"
                                                },
                                                "gt": {
                                                    "type": "string"
                                                },
                                                "gte": {
                                                    "type": "string"
                                                },
                                                "contains": {
                                                    "type": "string"
                                                },
                                                "startsWith": {
                                                    "type": "string"
                                                },
                                                "endsWith": {
                                                    "type": "string"
                                                },
                                                "mode": {},
                                                "not": {}
                                            }
                                        },
                                        "price_plan_id": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "equals": {
                                                    "type": "integer"
                                                },
                                                "in": {
                                                    "type": "array",
                                                    "items": {
                                                        "type": "integer"
                                                    }
                                                },
                                                "notIn": {
                                                    "type": "array",
                                                    "items": {
                                                        "type": "integer"
                                                    }
                                                },
                                                "lt": {
                                                    "type": "integer"
                                                },
                                                "lte": {
                                                    "type": "integer"
                                                },
                                                "gt": {
                                                    "type": "integer"
                                                },
                                                "gte": {
                                                    "type": "integer"
                                                },
                                                "not": {}
                                            }
                                        },
                                        "created_at": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "equals": {},
                                                "in": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "notIn": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "lt": {},
                                                "lte": {},
                                                "gt": {},
                                                "gte": {},
                                                "not": {}
                                            }
                                        },
                                        "updated_at": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "equals": {},
                                                "in": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "notIn": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "lt": {},
                                                "lte": {},
                                                "gt": {},
                                                "gte": {},
                                                "not": {}
                                            }
                                        },
                                        "price_plan": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "is": {},
                                                "isNot": {}
                                            }
                                        },
                                        "api": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "every": {},
                                                "some": {},
                                                "none": {}
                                            }
                                        },
                                        "environment": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "every": {},
                                                "some": {},
                                                "none": {}
                                            }
                                        },
                                        "namespace_members": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "every": {},
                                                "some": {},
                                                "none": {}
                                            }
                                        }
                                    }
                                }
                            },
                            "NOT": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "AND": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "price_plan_id": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "price_plan": {},
                                            "api": {},
                                            "environment": {},
                                            "namespace_members": {}
                                        }
                                    },
                                    "OR": {
                                        "type": "array",
                                        "items": {
                                            "additionalProperties": false,
                                            "type": "object",
                                            "properties": {
                                                "AND": {},
                                                "OR": {
                                                    "type": "array",
                                                    "items": {}
                                                },
                                                "NOT": {},
                                                "id": {},
                                                "name": {},
                                                "price_plan_id": {},
                                                "created_at": {},
                                                "updated_at": {},
                                                "price_plan": {},
                                                "api": {},
                                                "environment": {},
                                                "namespace_members": {}
                                            }
                                        }
                                    },
                                    "NOT": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "price_plan_id": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "price_plan": {},
                                            "api": {},
                                            "environment": {},
                                            "namespace_members": {}
                                        }
                                    },
                                    "id": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "mode": {},
                                            "not": {}
                                        }
                                    },
                                    "name": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "mode": {},
                                            "not": {}
                                        }
                                    },
                                    "price_plan_id": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "integer"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "integer"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "integer"
                                                }
                                            },
                                            "lt": {
                                                "type": "integer"
                                            },
                                            "lte": {
                                                "type": "integer"
                                            },
                                            "gt": {
                                                "type": "integer"
                                            },
                                            "gte": {
                                                "type": "integer"
                                            },
                                            "not": {}
                                        }
                                    },
                                    "created_at": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {},
                                            "in": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "lt": {},
                                            "lte": {},
                                            "gt": {},
                                            "gte": {},
                                            "not": {}
                                        }
                                    },
                                    "updated_at": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {},
                                            "in": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "lt": {},
                                            "lte": {},
                                            "gt": {},
                                            "gte": {},
                                            "not": {}
                                        }
                                    },
                                    "price_plan": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "is": {},
                                            "isNot": {}
                                        }
                                    },
                                    "api": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "every": {},
                                            "some": {},
                                            "none": {}
                                        }
                                    },
                                    "environment": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "every": {},
                                            "some": {},
                                            "none": {}
                                        }
                                    },
                                    "namespace_members": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "every": {},
                                            "some": {},
                                            "none": {}
                                        }
                                    }
                                }
                            },
                            "id": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "equals": {
                                        "type": "string"
                                    },
                                    "in": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "notIn": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "lt": {
                                        "type": "string"
                                    },
                                    "lte": {
                                        "type": "string"
                                    },
                                    "gt": {
                                        "type": "string"
                                    },
                                    "gte": {
                                        "type": "string"
                                    },
                                    "contains": {
                                        "type": "string"
                                    },
                                    "startsWith": {
                                        "type": "string"
                                    },
                                    "endsWith": {
                                        "type": "string"
                                    },
                                    "mode": {
                                        "type": "string",
                                        "enum": [
                                            "default",
                                            "insensitive"
                                        ]
                                    },
                                    "not": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "not": {}
                                        }
                                    }
                                }
                            },
                            "name": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "equals": {
                                        "type": "string"
                                    },
                                    "in": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "notIn": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "lt": {
                                        "type": "string"
                                    },
                                    "lte": {
                                        "type": "string"
                                    },
                                    "gt": {
                                        "type": "string"
                                    },
                                    "gte": {
                                        "type": "string"
                                    },
                                    "contains": {
                                        "type": "string"
                                    },
                                    "startsWith": {
                                        "type": "string"
                                    },
                                    "endsWith": {
                                        "type": "string"
                                    },
                                    "mode": {
                                        "type": "string",
                                        "enum": [
                                            "default",
                                            "insensitive"
                                        ]
                                    },
                                    "not": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "string"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "string"
                                                }
                                            },
                                            "lt": {
                                                "type": "string"
                                            },
                                            "lte": {
                                                "type": "string"
                                            },
                                            "gt": {
                                                "type": "string"
                                            },
                                            "gte": {
                                                "type": "string"
                                            },
                                            "contains": {
                                                "type": "string"
                                            },
                                            "startsWith": {
                                                "type": "string"
                                            },
                                            "endsWith": {
                                                "type": "string"
                                            },
                                            "not": {}
                                        }
                                    }
                                }
                            },
                            "price_plan_id": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "equals": {
                                        "type": "integer"
                                    },
                                    "in": {
                                        "type": "array",
                                        "items": {
                                            "type": "integer"
                                        }
                                    },
                                    "notIn": {
                                        "type": "array",
                                        "items": {
                                            "type": "integer"
                                        }
                                    },
                                    "lt": {
                                        "type": "integer"
                                    },
                                    "lte": {
                                        "type": "integer"
                                    },
                                    "gt": {
                                        "type": "integer"
                                    },
                                    "gte": {
                                        "type": "integer"
                                    },
                                    "not": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {
                                                "type": "integer"
                                            },
                                            "in": {
                                                "type": "array",
                                                "items": {
                                                    "type": "integer"
                                                }
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {
                                                    "type": "integer"
                                                }
                                            },
                                            "lt": {
                                                "type": "integer"
                                            },
                                            "lte": {
                                                "type": "integer"
                                            },
                                            "gt": {
                                                "type": "integer"
                                            },
                                            "gte": {
                                                "type": "integer"
                                            },
                                            "not": {}
                                        }
                                    }
                                }
                            },
                            "created_at": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "equals": {
                                        "type": "string"
                                    },
                                    "in": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "notIn": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "lt": {
                                        "type": "string"
                                    },
                                    "lte": {
                                        "type": "string"
                                    },
                                    "gt": {
                                        "type": "string"
                                    },
                                    "gte": {
                                        "type": "string"
                                    },
                                    "not": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {},
                                            "in": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "lt": {},
                                            "lte": {},
                                            "gt": {},
                                            "gte": {},
                                            "not": {}
                                        }
                                    }
                                }
                            },
                            "updated_at": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "equals": {
                                        "type": "string"
                                    },
                                    "in": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "notIn": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    },
                                    "lt": {
                                        "type": "string"
                                    },
                                    "lte": {
                                        "type": "string"
                                    },
                                    "gt": {
                                        "type": "string"
                                    },
                                    "gte": {
                                        "type": "string"
                                    },
                                    "not": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "equals": {},
                                            "in": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "notIn": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "lt": {},
                                            "lte": {},
                                            "gt": {},
                                            "gte": {},
                                            "not": {}
                                        }
                                    }
                                }
                            },
                            "price_plan": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "is": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "quota_daily_requests": {},
                                            "quota_environments": {},
                                            "quota_members": {},
                                            "quota_apis": {},
                                            "allow_secondary_environments": {},
                                            "namespace": {}
                                        }
                                    },
                                    "isNot": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "quota_daily_requests": {},
                                            "quota_environments": {},
                                            "quota_members": {},
                                            "quota_apis": {},
                                            "allow_secondary_environments": {},
                                            "namespace": {}
                                        }
                                    }
                                }
                            },
                            "api": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "every": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "namespace_id": {},
                                            "name": {},
                                            "markdown_description": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "namespace": {},
                                            "deployment": {}
                                        }
                                    },
                                    "some": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "namespace_id": {},
                                            "name": {},
                                            "markdown_description": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "namespace": {},
                                            "deployment": {}
                                        }
                                    },
                                    "none": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "namespace_id": {},
                                            "name": {},
                                            "markdown_description": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "namespace": {},
                                            "deployment": {}
                                        }
                                    }
                                }
                            },
                            "environment": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "every": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "namespace_id": {},
                                            "primary_hostname": {},
                                            "hostnames": {},
                                            "primary": {},
                                            "namespace": {},
                                            "deployment_environment": {},
                                            "environment_edges": {},
                                            "nodepool_environment": {}
                                        }
                                    },
                                    "some": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "namespace_id": {},
                                            "primary_hostname": {},
                                            "hostnames": {},
                                            "primary": {},
                                            "namespace": {},
                                            "deployment_environment": {},
                                            "environment_edges": {},
                                            "nodepool_environment": {}
                                        }
                                    },
                                    "none": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "id": {},
                                            "name": {},
                                            "namespace_id": {},
                                            "primary_hostname": {},
                                            "hostnames": {},
                                            "primary": {},
                                            "namespace": {},
                                            "deployment_environment": {},
                                            "environment_edges": {},
                                            "nodepool_environment": {}
                                        }
                                    }
                                }
                            },
                            "namespace_members": {
                                "additionalProperties": false,
                                "type": "object",
                                "properties": {
                                    "every": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "user_id": {},
                                            "namespace_id": {},
                                            "membership": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "namespace": {},
                                            "users": {}
                                        }
                                    },
                                    "some": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "user_id": {},
                                            "namespace_id": {},
                                            "membership": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "namespace": {},
                                            "users": {}
                                        }
                                    },
                                    "none": {
                                        "additionalProperties": false,
                                        "type": "object",
                                        "properties": {
                                            "AND": {},
                                            "OR": {
                                                "type": "array",
                                                "items": {}
                                            },
                                            "NOT": {},
                                            "user_id": {},
                                            "namespace_id": {},
                                            "membership": {},
                                            "created_at": {},
                                            "updated_at": {},
                                            "namespace": {},
                                            "users": {}
                                        }
                                    }
                                }
                            }
                        }
                    }
                },
                "additionalProperties": false
            },
        }
    ]);
});
const complexSchema = `
input CreateNamespace {
  name: String!
  personal: Boolean!
}

input NamespaceUpdateMembership {
  namespaceID: ID!
  memberID: ID!
  newMembership: Membership!
}

union DeleteApiResponse = Success | Error

type ApiUpdated {
  api: Api!
}

input DeleteDeployment {
  deploymentID: ID!
}

union AdminConfigResponse = Error | AdminConfig

input NamespaceRemoveMember {
  namespaceID: ID!
  memberID: ID!
}

type User {
  id: ID!
  name: String!
  email: String!
  namespaces: [Namespace!]!
  accessTokens: [AccessToken!]!
}

input DeleteApi {
  id: ID!
}

union NamespaceUpdateMembershipResponse = NamespaceMembershipUpdated | Error

type NamespaceMembershipUpdated {
  namespace: Namespace!
}

type EnvironmentUpdated {
  environment: Environment!
}

union CreateNamespaceResponse = NamespaceCreated | Error

type Namespace {
  id: ID!
  name: String!
  members: [Member!]!
  apis: [Api!]!
  environments: [Environment!]!
  personal: Boolean!
}

input UpdateEnvironment {
  environmentID: ID!
  edgeIDs: [ID!]
}

input DeleteEnvironment {
  environmentID: ID!
}

enum ApiVisibility {
  public
  private
  namespace
}

type Member {
  user: User!
  membership: Membership!
}

union DeleteAccessTokenResponse = Success | Error

input CreateApi {
  apiName: String!
  namespaceID: String!
  visibility: ApiVisibility!
  markdownDescription: String!
}

union CreateApiResponse = ApiCreated | Error

union UpdateApiResponse = ApiUpdated | Error

input CreateOrUpdateDeployment {
  apiID: ID!
  name: String
  config: JSON!
  environmentIDs: [ID!]!
}

union CreateOrUpdateDeploymentResponse = DeploymentCreated | DeploymentUpdated | Error

union CreateEnvironmentResponse = EnvironmentCreated | Error

input DeleteAccessToken {
  id: ID!
}

union CreateAccessTokenResponse = AccessTokenCreated | Error

input UpdateApi {
  id: ID!
  apiName: String!
  config: JSON!
  visibility: ApiVisibility!
  markdownDescription: String!
}

type Success {
  message: String!
}

scalar JSON

type AdminConfig {
  WunderNodeImageTag: String!
}

input NamespaceAddMember {
  namespaceID: ID!
  newMemberEmail: String!
  membership: Membership
}

input DeleteNamespace {
  namespaceID: ID!
}

type AccessTokenCreated {
  token: String!
  accessToken: AccessToken!
}

union NamespaceAddMemberResponse = NamespaceMemberAdded | Error

type Api {
  id: ID!
  name: String!
  visibility: ApiVisibility!
  deployments: [Deployment!]!
  markdownDescription: String!
}

union DeleteDeploymentResponse = Success | Error

type Query {
  user: User
  edges: [Edge!]!
  admin_Config: AdminConfigResponse!
}

type NamespaceMemberRemoved {
  namespace: Namespace!
}

type NamespaceMemberAdded {
  namespace: Namespace!
}

union DeleteNamespaceResponse = Success | Error

enum Membership {
  owner
  maintainer
  viewer
  guest
}

input CreateAccessToken {
  name: String!
}

union DeleteEnvironmentResponse = Success | Error

scalar Time

type ApiCreated {
  api: Api!
}

enum EnvironmentKind {
  Personal
  Team
  Business
}

type Edge {
  id: ID!
  name: String!
  location: String!
}

union NamespaceRemoveMemberResponse = NamespaceMemberRemoved | Error

type NamespaceCreated {
  namespace: Namespace!
}

union UpdateEnvironmentResponse = EnvironmentUpdated | Error

type Deployment {
  id: ID!
  name: String!
  config: JSON!
  environments: [Environment!]!
}

type Error {
  code: ErrorCode!
  message: String!
}

type Mutation {
  accessTokenCreate(input: CreateAccessToken!): CreateAccessTokenResponse!
  accessTokenDelete(input: DeleteAccessToken!): DeleteAccessTokenResponse!
  apiCreate(input: CreateApi!): CreateApiResponse!
  apiUpdate(input: UpdateApi!): UpdateApiResponse!
  apiDelete(input: DeleteApi!): DeleteApiResponse!
  deploymentCreateOrUpdate(input: CreateOrUpdateDeployment!): CreateOrUpdateDeploymentResponse!
  deploymentDelete(input: DeleteDeployment!): DeleteDeploymentResponse!
  environmentCreate(input: CreateEnvironment!): CreateEnvironmentResponse!
  environmentUpdate(input: UpdateEnvironment!): UpdateEnvironmentResponse!
  environmentDelete(input: DeleteEnvironment!): DeleteEnvironmentResponse!
  namespaceCreate(input: CreateNamespace!): CreateNamespaceResponse!
  namespaceDelete(input: DeleteNamespace!): DeleteNamespaceResponse!
  namespaceAddMember(input: NamespaceAddMember!): NamespaceAddMemberResponse!
  namespaceRemoveMember(input: NamespaceRemoveMember!): NamespaceRemoveMemberResponse!
  namespaceUpdateMembership(input: NamespaceUpdateMembership!): NamespaceUpdateMembershipResponse!
  admin_setWunderNodeImageTag(imageTag: String!): AdminConfigResponse!
}

type AccessToken {
  id: ID!
  name: String!
  createdAt: Time!
}

type EnvironmentCreated {
  environment: Environment!
}

type DeploymentUpdated {
  deployment: Deployment!
}

enum ErrorCode {
  Internal
  AuthenticationRequired
  Unauthorized
  NotFound
  Conflict
  UserAlreadyHasPersonalNamespace
  TeamPlanInPersonalNamespace
  InvalidName
  UnableToDeployEnvironment
  InvalidWunderGraphConfig
  ApiEnvironmentNamespaceMismatch
  UnableToUpdateEdgesOnPersonalEnvironment
}

type Environment {
  id: ID!
  name: String
  primary: Boolean!
  kind: EnvironmentKind!
  edges: [Edge!]
  primaryHostName: String!
  hostNames: [String!]!
}

type DeploymentCreated {
  deployment: Deployment!
}

input CreateEnvironment {
  namespace: ID!
  name: String
  primary: Boolean!
  kind: EnvironmentKind!
  edges: [ID!]
}`;
