// @wundergraph/sdk/dist/graphql/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cleanupSchema = void 0;
const graphql_1 = require("graphql");
const cleanupSchema = (schema) => {
    const printed = graphql_1.printSchema(schema, { commentDescriptions: true });
    const ast = graphql_1.parse(printed);
    const cleanAst = graphql_1.visit(ast, {
        FieldDefinition: {
            enter: node => {
                if (node.description) {
                    if (node.description.value === "") {
                        return {
                            ...node,
                            description: undefined,
                        };
                    }
                    if (node.description.block === true) {
                        const description = {
                            ...node.description,
                            block: false,
                        };
                        return {
                            ...node,
                            description,
                        };
                    }
                }
            }
        },
        InputValueDefinition: {
            enter: node => {
                if (node.description) {
                    if (node.description.value === "") {
                        return {
                            ...node,
                            description: undefined,
                        };
                    }
                    if (node.description.block === true) {
                        const description = {
                            ...node.description,
                            block: false,
                        };
                        return {
                            ...node,
                            description,
                        };
                    }
                }
            }
        }
    });
    return graphql_1.buildASTSchema(cleanAst);
};
exports.cleanupSchema = cleanupSchema;
