// @wundergraph/sdk/dist/linkbuilder/index.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("./index");
const graphql_1 = require("graphql");
const chai_1 = require("chai");
test("queryTypeFields", () => {
    const schema = graphql_1.buildSchema(`
        type Query {
            a: String!
            b(a: Int): Int!
            c(input: UserInput! a: String): User
        }
        type User {
            a: String
        }
        input UserInput {
            id: ID!
        }
    `);
    const actual = index_1.queryTypeFields(schema);
    const expected = [
        {
            typeName: "Query",
            fieldName: "a",
            fieldType: "String!",
            arguments: []
        },
        {
            typeName: "Query",
            fieldName: "b",
            fieldType: "Int!",
            arguments: [
                {
                    name: "a",
                    type: "Int"
                }
            ]
        },
        {
            typeName: "Query",
            fieldName: "c",
            fieldType: "User",
            arguments: [
                {
                    name: "input",
                    type: "UserInput!"
                },
                {
                    name: "a",
                    type: "String"
                }
            ]
        }
    ];
    chai_1.assert.equal(JSON.stringify(actual, null, "  "), JSON.stringify(expected, null, "  "));
});
test("linkBuilder", () => {
    const linkDefinition = index_1.sourceFieldStep()
        .source("user")
        .target("User", "userUser")
        .argument("age", "fieldArgument", "userAge")
        .argument("name", "objectField", "username")
        .build();
    console.log(JSON.stringify(linkDefinition, null, "  "));
});
