// @wundergraph/sdk/dist/codegen/templates/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = require("./typescript");
const client_1 = require("./typescript/client");
const react_1 = require("./typescript/react");
const mocks_1 = require("./typescript/mocks");
const namespaces_1 = require("./typescript/namespaces");
const apiclient_1 = require("../../apiclient");
const operations_1 = require("./typescript/operations");
const linkbuilder_1 = require("./typescript/linkbuilder");
const typescriptModels = [
    new typescript_1.TypeScriptInputModels(),
    new typescript_1.TypeScriptResponseModels(),
];
const typescriptAll = [
    ...typescriptModels,
    new client_1.TypeScriptClient(),
    new mocks_1.TypeScriptMocks(),
    new operations_1.Operations(),
];
const namespaces = (config) => {
    const client = new apiclient_1.ApiClient(config);
    return new namespaces_1.NamespacesTemplate(client);
};
const templates = {
    typescript: {
        // models generates all models for input and output
        models: typescriptModels,
        inputModels: new typescript_1.TypeScriptInputModels(),
        responseModels: new typescript_1.TypeScriptResponseModels(),
        client: new client_1.TypeScriptClient(),
        all: typescriptAll,
        mocks: new mocks_1.TypeScriptMocks(),
        react: [
            ...typescriptAll,
            new react_1.TypescriptReactProvider(),
            new react_1.TypescriptReactHooks(),
        ],
        namespaces: namespaces(),
        customNamespaces: namespaces,
        operations: new operations_1.Operations(),
        linkBuilder: new linkbuilder_1.TypeScriptLinkBuilder(),
    }
};
exports.default = templates;
