// @wundergraph/sdk/dist/codegen/templates/typescript/index.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_test_1 = require("../../index.test");
const index_1 = require("./index");
test("TypeScriptInputModels", async () => {
    const models = `
    export interface CreatePetInput {
        petInput: {
            id: string,
            name?: string
        }
    }
    `;
    const out = await index_test_1.RunTemplateTest(new index_1.TypeScriptInputModels());
    out.equals({
        "generated/models.ts": index_1.formatTypeScript(models),
    });
});
test("TypeScriptResponseModels", async () => {
    const models = `
    
    export interface GraphQLError {
        message: string;
        path?: ReadonlyArray<string | number>;
    }

    export interface MyReviewsResponse {
        data: {
            me?: {
               name?: string
               reviews?: {
                id: string
                body?: string
               }[] 
            }
        };
        errors?: ReadonlyArray<GraphQLError>;
    }
    
    export interface CreatePetResponse {
        data: {
            postPets?: {
                name?: string
            }
        };
        errors?: ReadonlyArray<GraphQLError>;
    }
    
    export interface NewPetsResponse {
        data: {
            newPets?: {
                name?: string
            }
        };
        errors?: ReadonlyArray<GraphQLError>;
    }
    `;
    const out = await index_test_1.RunTemplateTest(new index_1.TypeScriptResponseModels());
    out.equals({
        "generated/models.ts": index_1.formatTypeScript(models),
    });
});
