// @wundergraph/sdk/dist/codegen/templates/typescript/linkbuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeScriptLinkBuilder = void 0;
const linkbuilder_1 = require("../../../linkbuilder");
const graphql_1 = require("graphql");
const handlebars_1 = __importDefault(require("handlebars"));
const index_1 = require("./index");
// @ts-ignore
const linkbuilder_ts_handlebars_1 = __importDefault(require("./linkbuilder.ts.handlebars"));
class TypeScriptLinkBuilder {
    generate(config) {
        const input = index_1.loadTemplate(linkbuilder_ts_handlebars_1.default);
        const tmpl = handlebars_1.default.compile(input);
        const schema = graphql_1.buildSchema(config.application.EngineConfiguration.Schema);
        const fields = linkbuilder_1.queryTypeFields(schema);
        const types = linkbuilder_1.typesInfo(schema);
        const model = {
            fields: fields.map(field => ({
                name: field.fieldName,
                args: field.arguments.map(arg => arg.name),
            })),
            types: types.map(t => ({
                name: t.typeName,
                fields: t.fieldNames,
            }))
        };
        const content = tmpl(model);
        return Promise.resolve([
            {
                path: "linkbuilder.ts",
                content: index_1.formatTypeScript(content),
            }
        ]);
    }
}
exports.TypeScriptLinkBuilder = TypeScriptLinkBuilder;
