// @wundergraph/sdk/dist/codegen/templates/typescript/operations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operations = void 0;
const index_1 = require("./index");
const wundernode_config_1 = require("@wundergraph/protobuf/dist/wundernode_config");
const handlebars_1 = __importDefault(require("handlebars"));
// @ts-ignore
const operations_ts_handlebars_1 = __importDefault(require("./operations.ts.handlebars"));
class Operations {
    generate(config) {
        const model = {
            operations: config.application.Operations.map(op => ({
                name: op.Name,
                isMutation: op.OperationType === wundernode_config_1.OperationType.MUTATION,
                isQuery: op.OperationType === wundernode_config_1.OperationType.QUERY,
                isSubscription: op.OperationType === wundernode_config_1.OperationType.SUBSCRIPTION,
            }))
        };
        const input = index_1.loadTemplate(operations_ts_handlebars_1.default);
        const tmpl = handlebars_1.default.compile(input);
        const content = tmpl(model);
        return Promise.resolve([
            {
                path: "operations.ts",
                content: index_1.formatTypeScript(content)
            }
        ]);
    }
}
exports.Operations = Operations;
